source/extensions/filters/network/thrift_proxy/filter.cc

```diff
@@ -0,0 +1,307 @@
+                                             std::make_unique<AutoProtocolImpl>(req_callbacks_));
+  }
+
+  ENVOY_LOG(trace, "thrift: read {} bytes", data.length());
+  req_buffer_.move(data);
+
+  try {
+    BufferWrapper wrapped(req_buffer_);
+
+    req_decoder_->onData(wrapped);
+
+    // Move consumed portion of request back to data for the upstream to consume.
+    uint64_t pos = wrapped.position();
+    if (pos > 0) {
+      data.move(req_buffer_, pos);
+    }
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG(error, "thrift error: {}", ex.what());
+    req_decoder_.reset();
+    data.move(req_buffer_);
+    stats_.request_decoding_error_.inc();
+    sniffing_ = false;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus Filter::onWrite(Buffer::Instance& data, bool) {
+  if (!sniffing_) {
+    if (resp_buffer_.length() > 0) {
+      // Stopped sniffing during request (in onData). Make sure resp_buffer_ contents are at the
+      // start of data or the downstream will see a corrupted response.
+      resp_buffer_.move(data);
+      data.move(resp_buffer_);
+      ASSERT(resp_buffer_.length() == 0);
+    }
+
+    return Network::FilterStatus::Continue;
+  }
+
+  if (resp_decoder_ == nullptr) {
+    resp_decoder_ = std::make_unique<Decoder>(std::make_unique<AutoTransportImpl>(resp_callbacks_),
+                                              std::make_unique<AutoProtocolImpl>(resp_callbacks_));
+  }
+
+  ENVOY_LOG(trace, "thrift wrote {} bytes", data.length());
+  resp_buffer_.move(data);
+
+  try {
+    BufferWrapper wrapped(resp_buffer_);
+
+    resp_decoder_->onData(wrapped);
+
+    // Move consumed portion of response back to data for the downstream to consume.
+    uint64_t pos = wrapped.position();
+    if (pos > 0) {
+      data.move(resp_buffer_, pos);
+    }
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG(error, "thrift error: {}", ex.what());
+    resp_decoder_.reset();
+    data.move(resp_buffer_);
+
+    stats_.response_decoding_error_.inc();
+    sniffing_ = false;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Filter::chargeDownstreamRequestStart(MessageType msg_type, int32_t seq_id) {
+  if (req_ != nullptr) {
+    throw EnvoyException("unexpected request messageStart callback");
+  }
+
+  req_ = std::make_unique<ActiveMessage>(*this, msg_type, seq_id);
+
+  stats_.request_.inc();
+  switch (msg_type) {
+  case MessageType::Call:
+    stats_.request_call_.inc();
+    break;
+  case MessageType::Oneway:
+    stats_.request_oneway_.inc();
+    break;
+  default:
+    stats_.request_invalid_type_.inc();
+    break;
+  }
+}
+
+void Filter::chargeDownstreamRequestComplete() {
+  if (req_ == nullptr) {
+    throw EnvoyException("unexpected request messageComplete callback");
+  }
+
+  // One-way messages do not receive responses.
+  if (req_->msg_type_ == MessageType::Oneway) {
+    req_.reset();
+    return;
+  }
+
+  active_call_list_.emplace_back(std::move(req_));
```

Comment:
should we have a safety check and drop requests if we reach N=TOO_MANY active calls? I agree N=1 most of times, but I am not sure if we are covered against extreme failure cases (entire upstream cluster goes away or becomes overloaded)?

Reply:
I added a check at the start of new requests. I put in an arbitrary limit of 64 pending requests at which point the filter will get stop sniffing.